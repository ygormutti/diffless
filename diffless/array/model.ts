import { JSIN } from '../jsin';
import { DiffLevel, Document, Equals, Excerpt, Location, Weigh } from '../model';

export type Tokenizer = (document: Document) => Token[];

export interface ArrayDiffOptions<TExcerpt extends Excerpt> {
    level: DiffLevel;
    excerptMapper: ExcerptMapper<TExcerpt>;
    similarityThreshold: number;
    equals?: Equals<TExcerpt>;
    weigh?: Weigh<TExcerpt>;
}

export type ExcerptMapper<TExcerpt extends Excerpt> = (document: Document) => TExcerpt[];

@JSIN.enabled
export class Token extends Excerpt {
    constructor(
        location: Location,
        content: string,
        readonly type: number,
    ) {
        super(content, location);
    }

    equals(other: Token): boolean {
        return !(other instanceof ValuedToken) && this.type === other.type;
    }

    static equals(a: Token, b: Token) {
        return a.equals(b);
    }
}

@JSIN.enabled
export class ValuedToken<TTokenValue> extends Token {
    constructor(
        location: Location,
        content: string,
        type: number,
        readonly value: TTokenValue,
        readonly valueEquals: Equals<TTokenValue>,
    ) {
        super(location, content, type);
    }

    equals(other: Token): boolean {
        return other instanceof ValuedToken
            && this.type === other.type
            && this.valueEquals(this.value, other.value);
    }
}
