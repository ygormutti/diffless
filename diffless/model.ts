import { JSIN } from './jsin';

/**
 * Model classes and interfaces.
 *
 * Design and docs inspired by: https://microsoft.github.io/language-server-protocol/specification
 */

/**
 * The URI of a document
 */
export type DocumentURI = string;

/**
 * Function type that checks if two objects are equal according to some criteria
 */
export type Equals<T> = (a: T, b: T) => boolean;

/**
 * Function type that returns the object weigh
 */
export type Weigh<T> = (obj: T) => number;

/**
 * The end of line sequence adopted for internal use
 */
export const EOL = '\n';

/**
 * A type which instances have a natural total order
 */
export interface Comparable<T> {
    compareTo(other: T): number;
}

/**
 * Position in a text document expressed as one-based line and one-based character offset.
 *
 * A position is between two characters like an ‘insert’ cursor in a editor.
 */
@JSIN.enabled
export class Position implements Comparable<Position> {
    constructor(
        /**
         * Line position in a document (one-based).
         */
        readonly line: number,

        /**
         * Character offset on a line in a document (one-based). Assuming that the line is
         * represented as a string, the `character` value represents the gap between the
         * `character` and `character + 1`.
         *
         * If the character value is greater than the line length it defaults back to the
         * line length.
         */
        readonly character: number,
    ) { }

    get lineOffset() {
        return this.line - 1;
    }

    get characterOffset() {
        return this.character - 1;
    }

    get code() {
        return `${this.line}:${this.character}`;
    }

    equals(other: Position) {
        return this.compareTo(other) === 0;
    }

    compareTo(other: Position): number {
        let value = this.line - other.line;
        if (!value) { value = this.character - other.character; }
        return value;
    }
}

/**
 * A range in a text document expressed as (one-based) start and end positions.
 *
 * A range is comparable to a selection in an editor. Therefore the end position is exclusive.
 * If you want to specify a range that contains a line including the line ending character(s)
 * then use an end position denoting the start of the next line.
 */
@JSIN.enabled
export class Range implements Comparable<Range> {
    constructor(
        /**
         * The range's start position.
         */
        readonly start: Position,

        /**
         * The range's end position.
         */
        readonly end: Position,
    ) { }

    get code() {
        return `[${(this.start.code)}; ${this.end.code})`;
    }

    compareTo(other: Range): number {
        let value = this.start.compareTo(other.start);
        if (!value) { value = this.end.compareTo(other.end); }
        return value;
    }

    static forLine(line: number) {
        return new Range(new Position(line, 1), new Position(line + 1, 1));
    }

    static forCharacter(position: Position, content: string) {
        const end = content === EOL ?
            new Position(position.line + 1, 1) :
            new Position(position.line, position.character + 1);
        return new Range(position, end);
    }
}

@JSIN.enabled
export class Location {
    constructor(
        readonly uri: DocumentURI,
        readonly range: Range,
    ) { }

    get code() {
        return `${this.range.code} @ ${this.uri}`;
    }

    static forLine(documentURI: DocumentURI, line: number) {
        return new Location(documentURI, Range.forLine(line));
    }

    static forCharacter(documentURI: DocumentURI, position: Position, content: string) {
        return new Location(documentURI, Range.forCharacter(position, content));
    }
}

/**
 * The unit of comparison of an atomic diff tool
 */
@JSIN.enabled
export class Atom {
    constructor(
        readonly content: string,
        readonly location: Location,
    ) { }

    get start() {
        return this.location.range.start;
    }

    get end() {
        return this.location.range.end;
    }

    static contentLength(atom: Atom) {
        return atom.content.length;
    }

    static sameContent(a: Atom, b: Atom) {
        return a.content === b.content;
    }
}

/**
 * A line from a text document
 */
@JSIN.enabled
export class Line extends Atom {
    constructor(
        content: string,
        documentURI: DocumentURI,
        readonly line: number,
    ) {
        super(content, Location.forLine(documentURI, line));
    }
}

/**
 * A single character from a text document
 */
@JSIN.enabled
export class Character extends Atom {
    constructor(
        content: string,
        documentURI: DocumentURI,
        readonly position: Position,
    ) {
        super(content, Location.forCharacter(documentURI, position, content));
    }
}

/**
 * The default content type for text documents
 */
export const DEFAULT_CONTENT_TYPE = 'text/plain';

/**
 * A text document
 */
@JSIN.enabled
export class Document {
    readonly lines: Line[];
    readonly characters: Character[];

    constructor(
        readonly uri: DocumentURI,
        readonly content: string,
        readonly contentType: string = DEFAULT_CONTENT_TYPE,
    ) {
        const linesContents = normalizeEOLs(content).split(EOL);
        this.lines = buildLines(linesContents, uri);
        this.characters = buildCharacters(this.lines, uri);
    }

    getRange(range: Range): string {
        const { start, end } = range;
        const lines = this.lines.filter((_, i) => start.lineOffset <= i && i <= end.lineOffset);

        if (lines.length === 1) {
            return lines[0].content.substring(start.characterOffset, end.characterOffset);
        }

        let rangeString = '';
        for (let lineOffset = 0; lineOffset < lines.length; lineOffset++) {
            const line = lines[lineOffset].content;
            let startOffset = 0;
            let endOffset = line.length;
            if (lineOffset === 0) {
                // first line
                startOffset = start.characterOffset;
            } else if (lineOffset === lines.length - 1) {
                // last line
                endOffset = end.characterOffset;
            }

            rangeString += line.substring(startOffset, Math.min(line.length, endOffset));
        }
        return rangeString;
    }
}

function buildLines(linesContents: string[], documentURI: DocumentURI): Line[] {
    return linesContents.map((v, i) => new Line(v + EOL, documentURI, i + 1));
}

function buildCharacters(lines: Line[], documentURI: DocumentURI): Character[] {
    const characters: Character[] = [];
    for (const [lineOffset, line] of lines.entries()) {
        const lineNumber = lineOffset + 1;
        for (const [characterOffset, character] of Array.from(line.content).entries()) {
            const position = new Position(lineNumber, characterOffset + 1);
            characters.push(new Character(character, documentURI, position));
        }
    }
    return characters;
}

function normalizeEOLs(content: string): string {
    return content.replace('\r\n', EOL).replace('\r', EOL);
}

export enum EditOperation {
    Add,
    Delete,
    Replace,
    Move,
    Copy,
    Rename,
}

export enum DiffLevel {
    Binary,
    Textual,
    Lexical,
    Syntactic,
    Semantic,
    DataFlow,
}

@JSIN.enabled
export class DiffItem {
    constructor(
        readonly level: DiffLevel,
        readonly left?: Location,
        readonly right?: Location,
    ) { }
}

@JSIN.enabled
export class Edit extends DiffItem implements Comparable<Edit> {
    constructor(
        level: DiffLevel,
        readonly operation: EditOperation,
        left?: Location,
        right?: Location,
    ) {
        super(level, left, right);
    }

    get code() {
        return `${DiffLevel[this.level]} ${EditOperation[this.operation].toLowerCase()}, ` +
            (this.left ? `L: ${this.left!.code}` : '') +
            (this.left && this.right ? ', ' : '') +
            (this.right ? `R: ${this.right!.code}` : '');
    }

    compareTo(other: Edit): number {
        let value = other.level - this.level;
        if (!value) { value = other.operation - this.operation; }
        if (!value && this.right && other.right) { value = this.right.range.compareTo(other.right.range); }
        if (!value && this.left && other.left) { value = this.left.range.compareTo(other.left.range); }
        return value;
    }
}

@JSIN.enabled
export class Similarity extends DiffItem { }

@JSIN.enabled
export class DocumentDiff {
    constructor(
        readonly left: Document,
        readonly right: Document,
        readonly edits: Edit[],
        readonly similarities: Similarity[],
    ) { }
}

export type DiffTool = (left: Document, right: Document) => DocumentDiff;

export type DiffToolFactory = (options: any) => DiffTool;
