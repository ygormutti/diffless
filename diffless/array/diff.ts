import {
    DiffLevel,
    Document,
    DocumentDiff,
    Edit,
    EditType,
    Equal,
    Excerpt,
    Location,
    Position,
    Range,
} from '../model';
import { LCS, LCSResult } from './lcs';

export interface ArrayDiffOptions<TItem extends Excerpt> {
    level: DiffLevel;
    equal: Equal<TItem>;
    itemMapper: ItemMapper<TItem>;
    lcsThreshold: number;
    lcs: LCS;
}

export type ItemMapper<TItem extends Excerpt> = (document: Document) => TItem[];

class ItemWrapper<TItem extends Excerpt> {
    constructor(
        readonly item: TItem,
        public paired: boolean = false,
    ) { }
}

export function arrayDiff<TItem extends Excerpt>(
    options: ArrayDiffOptions<TItem>,
    left: Document,
    right: Document,
): DocumentDiff {
    const { level, equal, itemMapper, lcsThreshold, lcs } = options;
    const leftWrapped = itemMapper(left).map(wrapItem);
    const rightWrapped = itemMapper(right).map(wrapItem);
    const equalWrapped = wrapEqual(equal);
    const lcsResults: LCSResult<ItemWrapper<TItem>>[] = [];

    let result = lcs(equalWrapped, leftWrapped, rightWrapped);
    while (result.lcs.length > lcsThreshold) {
        pairWrappers(leftWrapped, result.leftOffset, result.lcs.length);
        pairWrappers(rightWrapped, result.rightOffset, result.lcs.length);
        lcsResults.push(result);
        result = lcs(equalWrapped, leftWrapped, rightWrapped);
    }

    let edits: Edit[] = [];

    const deletions = findUnpairedRanges(leftWrapped).map(
        r => new Edit(level, EditType.Delete, new Location(left.uri, r), undefined),
    );
    edits = edits.concat(deletions);

    const additions = findUnpairedRanges(rightWrapped).map(
        r => new Edit(level, EditType.Add, undefined, new Location(right.uri, r)),
    );
    edits = edits.concat(additions);

    const moves = findMoves(lcsResults, leftWrapped, rightWrapped, left, right, level);
    edits = edits.concat(moves);

    return new DocumentDiff(left, right, edits, []);
}

function wrapItem<TItem extends Excerpt>(item: TItem): ItemWrapper<TItem> {
    return new ItemWrapper(item);
}

function wrapEqual<TItem extends Excerpt>(equal: Equal<TItem>): Equal<ItemWrapper<TItem>> {
    return (l: ItemWrapper<TItem>, r: ItemWrapper<TItem>): boolean => {
        return !r.paired && !l.paired && equal(l.item, r.item);
    };
}

function pairWrappers(array: ItemWrapper<Excerpt>[], offset: number, length: number) {
    for (let i = offset; i < (offset + length); i++) {
        const item = array[i];
        item.paired = true;
    }
}

function findUnpairedRanges(array: ItemWrapper<Excerpt>[]): Range[] {
    const ranges = [];

    let start: Position | undefined;
    let end: Position | undefined;
    for (const wrapper of array) {
        const { paired, item: { range } } = wrapper;
        end = range.start;
        if (start === undefined) {
            if (paired) {
                continue;
            } else {
                start = range.start;
            }
        } else {
            if (paired) {
                ranges.push(new Range(start, end));
                start = undefined;
            }
        }
    }

    if (start !== undefined && end !== undefined && !start.equals(end)) {
        ranges.push(new Range(start, end));
    }

    return ranges;
}

function findMoves(
    lcsResults: LCSResult<ItemWrapper<Excerpt>>[],
    leftWrapped: ItemWrapper<Excerpt>[],
    rightWrapped: ItemWrapper<Excerpt>[],
    left: Document,
    right: Document,
    level: DiffLevel,
): Edit[] {
    const edits: Edit[] = [];

    const rightSortedResults = lcsResults.slice().sort((a, b) => a.rightOffset - b.rightOffset);
    const leftSortedResults = lcsResults.slice().sort((a, b) => a.leftOffset - b.leftOffset);
    for (let l = 0, r = 0; l < lcsResults.length && r < lcsResults.length;) {
        const leftResult = leftSortedResults[l];
        const rightResult = rightSortedResults[r];

        if (leftResult === rightResult) {
            l++;
            r++;
        } else {
            const rightLongerThanLeft = rightResult.lcs.length >= leftResult.lcs.length;
            // heuristic to favor cheaper edit result, then moves in right side
            const result = rightLongerThanLeft ? leftResult : rightResult;
            const { lcs: { length }, leftOffset, rightOffset } = result;

            const leftRange = buildMoveRange(leftWrapped, leftOffset, length);
            const leftLocation = new Location(left.uri, leftRange);

            const rightRange = buildMoveRange(rightWrapped, rightOffset, length);
            const rightLocation = new Location(right.uri, rightRange);

            const edit = new Edit(level, EditType.Move, leftLocation, rightLocation);
            edits.push(edit);

            if (result === leftResult) {
                l++;
            } else {
                r++;
            }
        }
    }

    return edits;
}

function buildMoveRange(items: ItemWrapper<Excerpt>[], offset: number, length: number): Range {
    const start = items[offset].item.range.start;
    const end = items[offset + length - 1].item.range.end;
    return new Range(start, end);
}
