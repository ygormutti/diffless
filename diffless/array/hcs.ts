import { Equals, Weigh } from '../model';

/**
 * Heaviest Common Subsequence model and implementation
 */

/**
 * Function type that finds the HCS between two arrays
 */
export type HCS = <TItem>(
    equal: Equals<TItem>,
    weigh: Weigh<TItem>,
    left: TItem[],
    right: TItem[],
) => HCSResult<TItem>;

/**
 * The Heaviest Common Subsequence of two arrays with items from the respective arrays
 */
export class HCSResult<TItem> {
    constructor(
        readonly weight: number,
        readonly leftHCS: TItem[],
        readonly rightHCS: TItem[],
    ) { }

    get length() {
        return this.leftHCS.length;
    }
}

class HCSCell<TItem> {
    constructor(
        readonly weight: number = 0,
        readonly hcs: [TItem, TItem][] = [],
    ) { }
}

/**
 * Solves HCS with quadratic time and linear space complexity
 */
export function dynamicProgrammingHCS<TItem>(
    equals: Equals<TItem>,
    weigh: Weigh<TItem>,
    left: TItem[],
    right: TItem[],
): HCSResult<TItem> {
    if (!left || !left.length || !right || !right.length) {
        return new HCSResult(0, [], []);
    }

    const width = left.length + 1;
    const height = right.length + 1;

    let prevRowArray: HCSCell<TItem>[] = new Array(width);
    prevRowArray.fill(new HCSCell());

    for (let row = 1; row < height; row++) {
        const rowArray: HCSCell<TItem>[] = new Array(width);
        rowArray[0] = new HCSCell();

        for (let col = 1; col < width; col++) {
            const leftItem = left[col - 1];
            const rightItem = right[row - 1];
            if (equals(leftItem, rightItem)) {
                const prevCell = prevRowArray[col - 1];
                const newWeight = prevCell.weight + weigh(leftItem);
                rowArray[col] = new HCSCell(newWeight, [...prevCell.hcs, [leftItem, rightItem]]);
            } else {
                const previousCellCandidates = [rowArray[col - 1], prevRowArray[col]];
                previousCellCandidates.sort((a, b) => {
                    if (a.weight !== b.weight) {
                        return b.weight - a.weight;
                    }
                    return a.hcs.length - b.hcs.length;
                });
                rowArray[col] = previousCellCandidates[0];
            }
        }
        prevRowArray = rowArray;
    }

    const lastCell = prevRowArray[prevRowArray.length - 1];
    const { hcs, weight } = lastCell;
    const leftHCS = hcs.map(p => p[0]);
    const rightHCS = hcs.map(p => p[1]);
    return new HCSResult(weight, leftHCS, rightHCS);
}
