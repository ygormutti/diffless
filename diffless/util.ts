import { Edit, Location, Position } from './model';

export class EditIndex {
    private index: Edit[][][];

    constructor() {
        this.index = [];
    }

    add(location: Location, change: Edit) {
        const { start, end } = location.range;
        this.addToPosition(start, change);
        this.addToPosition(end, change);
    }

    private addToPosition(position: Position, change: Edit) {
        const { line, character } = position;
        if (!this.index[line]) { this.index[line] = []; }

        let changes = this.index[line][character];
        if (!changes) {
            changes = [];
            this.index[line][character] = changes;
        }

        changes.push(change);

        // sort by level
        changes.sort((a, b) => {
            return ((a.level * 100) - (b.level * 100)) + (a.type - b.type);
        });
    }

    get(position: Position): Edit[] {
        const { line, character } = position;
        if (!this.index[line] || !this.index[line][character]) { return []; }
        return this.index[line][character];
    }
}

export function buildEditIndexes(changes: Edit[]): [EditIndex, EditIndex] {
    const leftChangeIndex = new EditIndex();
    const rightChangeIndex = new EditIndex();
    changes.forEach(change => {
        const { left, right } = change;
        if (left) { leftChangeIndex.add(left, change); }
        if (right) { rightChangeIndex.add(right, change); }
    });
    return [leftChangeIndex, rightChangeIndex];
}

export function sortEdits(changes: Edit[]): void {
    changes.sort((a: Edit, b: Edit) => {
        let value = a.level - b.level;
        if (!value) {
            value = a.type - b.type;
        }
        return value;
    });
}

export function intEnumKeys(E: any): string[] {
    return Object.keys(E).filter(k => typeof E[k as any] === 'number');
}

/**
 * Allows Scala-like stripMargin
 *
 * Parameters are applied implicitly via ES2015.
 *
 * @example
 * // returns "The Number is:\n    100\nThanks for playing!"
 * let num = 100
 * let result = stripMargin`The Number is:
 *         |    ${num}
 *         |Thanks for playing!`
 * get the gist: https://gist.github.com/jimschubert/06fea56a6d2a1e7fdbc2
 */
export function stripMargin(template: TemplateStringsArray, ...expressions: any[]) {
    const result = template.reduce((accumulator, part, i) => {
        return accumulator + expressions[i - 1] + part;
    });

    return result.replace(/(\n|\r|\r\n)\s*\|/g, '$1');
}
