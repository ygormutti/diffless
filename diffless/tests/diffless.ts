import { ArrayDiffTool } from '../array/diff';
import { characterDiff, lineDiff } from '../index';
import { DiffLevel, Document } from '../model';
import { announceHtml } from '../tests/test-util';
import { stripMargin } from '../util';

describe('array/diff', () => {
    describe('charactersDiff', () => {
        it('should detect additions and deletions correctly', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |d
                |efg`,
            );
            const right = new Document('string:right', stripMargin
                `bc
                |de
                |fgh`,
            );

            const diff = characterDiff(left, right);
            expect(diff.edits).toMatchSnapshot();
            expect(diff.similarities).toMatchSnapshot();
            announceHtml(left, right, diff.edits, 'basic');
        });

        it('should detect moves correctly 1', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |def`,
            );
            const right = new Document('string:right', stripMargin
                `def
                |abc`,
            );

            const diff = characterDiff(left, right);
            expect(diff.edits).toMatchSnapshot();
            announceHtml(left, right, diff.edits, 'moves1');
        });

        it('should detect moves correctly 2', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |def
                |ghi`,
            );
            const right = new Document('string:right', stripMargin
                `def
                |gabc
                |ghi`,
            );

            const diff = characterDiff(left, right);
            expect(diff.edits).toMatchSnapshot();
            announceHtml(left, right, diff.edits, 'moves2');
        });

        it('should detect moves correctly 3', () => {
            const left = new Document('string:left', stripMargin
                `abc
                |def
                |ghi`,
            );
            const right = new Document('string:right', stripMargin
                `abc
                |ghi
                |def`,
            );

            const diff = characterDiff(left, right);
            expect(diff.edits).toMatchSnapshot();
            announceHtml(left, right, diff.edits, 'moves3');
        });
    });

    describe('lineDiff', () => {
        const left = new Document('string:left', stripMargin
            `abc
            |def
            |ghijklmn`,
        );
        const right = new Document('string:right', stripMargin
            `abc
            |ghijklmn
            |def`,
        );

        it('should consider line weight', () => {
            const diff = lineDiff(left, right);
            expect(diff.edits).toMatchSnapshot();
            announceHtml(left, right, diff.edits, 'move_lines_weighed');
        });

        it('should be possible to use no weight', () => {
            const tool = new ArrayDiffTool({
                excerptMapper: d => d.lines,
                level: DiffLevel.Textual,
                similarityThreshold: 0,
                weigh: _ => 1,
            });
            const diff = tool.run(left, right);
            expect(diff.edits).toMatchSnapshot();
            announceHtml(left, right, diff.edits, 'move_lines_not_weighed');
        });
    });
});
