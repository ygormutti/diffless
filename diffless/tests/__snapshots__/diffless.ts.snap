// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array/diff charactersDiff should detect additions and deletions correctly 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [2:2; 3:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [3:3; 4:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [1:1; 1:2),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [2:2; 3:2),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [3:4; 4:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
]
`;

exports[`array/diff charactersDiff should detect additions and deletions correctly 2`] = `
Array [
  Similarity {
    "left": Location {
      "range": [1:2; 2:2),
      "uri": "string:left",
    },
    "level": 1,
    "right": Location {
      "range": [1:1; 2:2),
      "uri": "string:right",
    },
  },
  Similarity {
    "left": Location {
      "range": [3:2; 3:4),
      "uri": "string:left",
    },
    "level": 1,
    "right": Location {
      "range": [3:1; 3:3),
      "uri": "string:right",
    },
  },
]
`;

exports[`array/diff charactersDiff should detect moves correctly 1 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [1:4; 2:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [2:4; 3:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [1:4; 2:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [2:4; 3:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [2:1; 2:4),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [1:1; 1:4),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array/diff charactersDiff should detect moves correctly 2 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [1:4; 2:2),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [1:4; 2:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [2:1; 2:4),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [1:1; 1:4),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array/diff charactersDiff should detect moves correctly 3 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [3:4; 4:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [3:4; 4:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [2:4; 3:4),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [1:4; 2:4),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array/diff lineDiff should be able to compare strings directly 1`] = `
Array [
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [1:1; 2:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": undefined,
    "level": "Textual",
    "right": Location {
      "range": [7:1; 8:1),
      "uri": "string:right",
    },
    "type": "Add",
  },
  Object {
    "left": Location {
      "range": [2:1; 3:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [4:1; 5:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": undefined,
    "type": "Delete",
  },
  Object {
    "left": Location {
      "range": [3:1; 4:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [2:1; 3:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
  Object {
    "left": Location {
      "range": [7:1; 8:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [4:1; 5:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array/diff lineDiff should be possible to use no weight 1`] = `
Array [
  Object {
    "left": Location {
      "range": [3:1; 4:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [2:1; 3:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;

exports[`array/diff lineDiff should consider line weight 1`] = `
Array [
  Object {
    "left": Location {
      "range": [2:1; 3:1),
      "uri": "string:left",
    },
    "level": "Textual",
    "right": Location {
      "range": [3:1; 4:1),
      "uri": "string:right",
    },
    "type": "Move",
  },
]
`;
